proposed/2022/vulnerabilities-in-restore.md

```diff
@@ -0,0 +1,391 @@
+
+#### Technical background
+
+To best understand the technical proposal, it's important to understand the [NuGet package installation process](https://docs.microsoft.com/en-us/nuget/concepts/package-installation-process).
+
+When a package is seen, the client talks to sources and downloads whichever versions are needed. When talking to V3 sources, the client use the `package content` [resource](https://docs.microsoft.com/en-us/nuget/api/package-base-address-resource), which is a static resource that only has information about the available package versions. Only what is contained *within* the package itself affect restore. Nothing else. This allows an equivalence between `http` and `local` sources.
+
+#### Dedicated vulnerability information resource
+
+Checking vulnerability will always take a performance hit, but we want to minimize that. We can:
+
+- Make fewer http calls.
+- Minimize the amount of data we are handling.
+
+Both of these can be achieved by adding a means to understand whether a source supports vulnerability. We'd add a resource such as the following:
+
+```json
+    {
+      "@id": "https://nikolchevulntest.blob.core.windows.net/newcontainer/vulnerabilityindex.json",
+      "@type": "VulnerabilityInfo/6.3.0",
+      "comment": "The endpoint for discovering information about vulnerabilities of packages in this package source."
+    },
+```
+
+When NuGet restores, we check whether a source supports vulnerabilities and if so it can go through a protocol of getting this information.
+
+The client needs all the data to make decisions about vulnerability. In order to not have the client redownload a file and limit the total number of HTTP calls we will ever make, the vulnerability resource will be an index that can contain multiple pages.
+
+- The vulnernability resource **must** be an array of objects.
+- Each object **must** contain:
+  - `@name`, a user friendly name for the page, **must** be unique.
+  - `@id`, the url that contains the data.
+  - `@updated`, a UTC timestamp when the content at `@id` was updated last.
+  - `comment`, a user friendly description.
+- The vulnerability resource **must** at minimum contain 1 page.
+- The pages within the vulnerability resource **must** be exclusive.
+- The vulnerability resource may 2 pages.
+  - One page, `base`, represents the data up to a certain point in time. This page **must** be updated periodically, based on a trigger.
+  - The second page, `update`, represents the data from the last update of `base`, to present. This page **should** be updated with new data.
+  - Periodically, for example once a month, the data from `update` **should** be migrated to `base`.
+  - If an entry needs to be removed from `base`, `base` should be updated.
+- Other partioning strategies are workable as well, as long as only 2 pages are receiving all the new data.  

```

Comment:
Are the `base` and `update` names required if using other partitioning strategies?

Reply:
Base and update aren't required names, but highly recommended.

The idea with 2 pages updating is that the client will always want to have all the data be up to date. 
It would allow refreshing to be super fast without an overly complicated protocol, as we'd only download the out of date files.

It's balance between CI (where the data is always download) and local, where the data is refreshed.

Reply:
The 2 page idea makes perfect sense. I'm more curious as to why we *require* a maximum of 2 pages; that seems odd.

Reply:
It's not really suggesting maximum of 2 pages.
Or at least, it's not supposed to :) 

What that line is trying to communicate is that as few pages as possible should be updated, so that the client is not forced to redownload pages all the time.

Reply:
I updated the spec to try to make it clearer that this is our suggestion, but servers can implement the data partitioning in different ways. The client will just look at the list of files, and their update timestamps to determine whether it can use a previously cached download or not.